keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,215 @@
+- https://github.com/kubernetes/kubernetes/issues/6979
+- https://github.com/kubernetes/kubernetes/issues/33766
+- https://github.com/kubernetes/kubernetes/issues/24198
+- https://github.com/kubernetes/kubernetes/issues/60340
+
+And then, for other people:
+- https://github.com/rancher/rancher/issues/13584
+- https://github.com/helm/charts/pull/12319
+- https://github.com/EnMasseProject/enmasse/pull/1974
+- https://github.com/helm/charts/pull/11546
+- https://github.com/kubernetes/kubernetes/pull/35343
+
+This is replacing a lot of previous work and long-standing effort:
+- Initially: https://github.com/kubernetes/community/issues/229, then
+- https://github.com/kubernetes/community/pull/278
+- https://github.com/kubernetes/community/pull/620
+- https://github.com/kubernetes/kubernetes/pull/44597
+- https://github.com/kubernetes/kubernetes/pull/50296
+- https://github.com/kubernetes/kubernetes/pull/70436
+
+Server-side [apply](http://features.k8s.io/555) is what enables this proposal to
+become possible.
+
+### Goals
+
+The goal is to enable a union or "oneof" semantic in Kubernetes types, both for
+in-tree types and for CRDs.
+
+### Non-Goals
+
+We're not planning to use this KEP to release the feature, but mostly as a way
+to document what we're doing.
+
+## Proposal
+
+In order to support unions in a backward compatible way in kubernetes, we're
+proposing the following changes.
+
+Note that this proposes unions to be "at most one of". Whether exactly one is
+supported or not should be implemented by the validation logic.
+
+### Go tags
+
+We're proposing a new type of tags for go types (in-tree types, and also
+kubebuilder types):
+
+- `// +union` before a field means that it is part of the union. Multiple fields
+  can have this prefix. These fields MUST BE optional, omitempty and pointers.
+- `// +discriminator` before a field means that this field is the discriminator
+  for the union. Only one field per structure can have this prefix. This field HAS
+  TO be a string, and CAN be optional.
+
+Note that this means that only one "union" type can be defined per structure,
+which is a deliberate decision. Go embedding which, once flattened, would define
+multiple unions in the same structure are also forbidden. Also note that we have
+no occurrences of these existing in core-types (AFAIK).
+
+### OpenAPI
+
+OpenAPI v3 already allows a "oneOf" form, which is accepted by CRD validation
+(and will continue to be accepted in the future). That oneOf form will be used
+for validation, but is "on-top" of this proposal (validation is considered as
+not included in this proposal).
+
+New extensions are created in the openapi to describe the behavior:
+- `x-kubernetes-union-discriminator: <discriminator>` is set to the name of the
+  discriminator field, if present,
+- `x-kubernetes-oneof-fields-discriminated: {"<fieldName>":
+  "<discriminatedName>"}` is a map of fields that belong to the union to their
```

Comment:
if we restrict us to one union per object, let's simplify this tag to a simple list of field names.

Also I would suggest a simpler tag name: x-kubernetes-union-fields

Reply:
it's not a list on purpose, because we really want to be able to have a discriminated name that is not explicitely the name of the field. For example today, we have an implicit discriminator in `DeploymentStrategyType` whose value must be `RolloutStrategy` (matching the field name `rolloutStrategy`). For volumes, it's not as simple as capitalizing, e.g. nfs -> NFS. Keeping this mapping of field name to discriminated name is useful.

Reply:
How about instead of "discriminated", we call this "union member", where the discriminator field value is the name of a union member?

Then could we have `x-kubernetes-union-member-fields: { "<fieldName>": "<unionMemberName>"}` ? Or, if I'm following @sttts, maybe we could have `{ "<unionMemberName>": "<listOfFieldNames>"}`